File: web/src/api/promptSettings.ts
import { requestJson } from './client'

export type PromptSettingKey = 'agentBehavior' | 'incidentResponse' | 'knowledgeExtraction' | 'memoryUsage' | 'organizationRules'
export type PromptOverrides = Record<PromptSettingKey, string>

export type PromptSettings = {
  schemaVersion: number
  revision: number
  overrides: PromptOverrides
  defaults: PromptOverrides
  effective: PromptOverrides
  updatedAt: string | null
  immutableSafetySummary: string
  maxPromptChars: number
}

export function getPromptSettings(): Promise<PromptSettings> {
  return requestJson<PromptSettings>('/api/prompt-settings')
}

export function updatePromptSettings(revision: number, overrides: PromptOverrides): Promise<PromptSettings> {
  return requestJson<PromptSettings>('/api/prompt-settings', {
    method: 'PUT',
    body: JSON.stringify({ revision, overrides }),
  })
}

export function resetPromptSettings(revision: number): Promise<PromptSettings> {
  return requestJson<PromptSettings>('/api/prompt-settings/reset', {
    method: 'POST',
    body: JSON.stringify({ revision }),
  })
}

File: src/app/api/schemas/prompt_settings.py
from pydantic import BaseModel, ConfigDict, Field


def _to_camel(value: str) -> str:
    first, *rest = value.split("_")
    return first + "".join(part.capitalize() for part in rest)


class PromptOverrides(BaseModel):
    agent_behavior: str = ""
    incident_response: str = ""
    knowledge_extraction: str = ""
    memory_usage: str = ""
    organization_rules: str = ""

    model_config = ConfigDict(alias_generator=_to_camel, populate_by_name=True, extra="forbid")


class PromptSettingsUpdateRequest(BaseModel):
    revision: int = Field(ge=0)
    overrides: PromptOverrides


class PromptSettingsResetRequest(BaseModel):
    revision: int = Field(ge=0)


class PromptSettingsView(BaseModel):
    schema_version: int
    revision: int
    overrides: PromptOverrides
    defaults: PromptOverrides
    effective: PromptOverrides
    updated_at: str | None
    immutable_safety_summary: str
    max_prompt_chars: int

    model_config = ConfigDict(alias_generator=_to_camel, populate_by_name=True)
